contracts/contracts/EvolvingRegistry.sol

```diff
@@ -0,0 +1,85 @@
+pragma solidity 0.4.23;
+
+import '../../node_modules/openzeppelin-solidity/contracts/ownership/Ownable.sol';
+
+contract EvolvingRegistry is Ownable {
+  event NewEntryType(uint16 indexed _entryTypeIndex);
+  event EntryTypeUpdate(uint16 indexed _entryTypeIndex);
+  event NewEntry(uint256 indexed _entryIndex, uint16 _entryTypeIndex);
+
+  struct EntryType {
+    address contractAddress;
+    string name;
+    bool isEnabled;
+  }
+
+  EntryType[2**16] public entryTypes;
+  uint16 entryTypeLength;
+
+  // Enables easy lookup of entry type by contract address instead of index
+  mapping(address => uint16) public entryTypesByAddress;
+
+  /* Entries, represented by an array of entry type indices.
+    The index of each entry is the entry id;
+    The value of each entry is the index of the entry type.
+    The index can be used to get the entry type contract.
+    The entry type contract handles all logic for the entry.
+  */
+  uint16[] public entries;
+
+  function addEntryType(address _contractAddress, string _name) public onlyOwner() {
+    // Don't allow the same contract address to be added twice
+    uint16 trialIndex = entryTypesByAddress[_contractAddress];
+    require(entryTypes[trialIndex].contractAddress != _contractAddress);
```

Comment:
Maybe add `require(_contractAddress != 0)`? `addEntry()` seems to depend on `entryTypesByAddress[0]` being 0.